src/command_modules/azure-cli-cloud/azure/cli/command_modules/cloud/custom.py

```diff
@@ -32,6 +34,33 @@ def show_cloud(cloud_name=None):
         raise CLIError(e)
 
 
+def _populate_from_metadata_endpoint(cloud, arm_endpoint):
+    endpoints_in_metadata = ['gallery', 'active_directory_graph_resource_id',
+                             'active_directory_resource_id', 'active_directory']
+    if not arm_endpoint or all([cloud.endpoints.has_endpoint_set(n) for n in endpoints_in_metadata]):
+        return
+    try:
+        import requests
+        metadata_endpoint = arm_endpoint + METADATA_ENDPOINT_SUFFIX
+        response = requests.get(metadata_endpoint)
+        if response.status_code == 200:
+            metadata = response.json()
+            if not cloud.endpoints.has_endpoint_set('gallery'):
+                setattr(cloud.endpoints, 'gallery', metadata.get('galleryEndpoint'))
+            if not cloud.endpoints.has_endpoint_set('active_directory_graph_resource_id'):
+                setattr(cloud.endpoints, 'active_directory_graph_resource_id', metadata.get('graphEndpoint'))
+            if not cloud.endpoints.has_endpoint_set('active_directory'):
+                setattr(cloud.endpoints, 'active_directory', metadata['authentication'].get('loginEndpoint'))
+            if not cloud.endpoints.has_endpoint_set('active_directory_resource_id'):
+                setattr(cloud.endpoints, 'active_directory_resource_id', metadata['authentication']['audiences'][0])
+        else:
+            raise CLIError('Server returned status code {} for {}'.format(response.status_code, metadata_endpoint))
+    except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError) as err:
+        raise CLIError('Please ensure you have network connection. Error detail: {}'.format(str(err)))
+    except ValueError as err:
+        raise CLIError('Response body does not contain valid json. Error detail: {}'.format(str(err)))
```

Comment:
I assume this and `ConnectionError` above will happen rather rarely unless server outage or users have no network? If yes, I suggest we don't do any polishing, just let it surface up. Afaik, request itself reports pretty clear messages when network is unavailable.